File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)
        prefixSum = [1 for _ in range(size)]
        suffixSum = [1 for _ in range(size)]
        ans = []

        for i in range(1, size):
            prefixSum[i] = nums[i - 1] * prefixSum[i - 1]
        for i in range(size - 2, -1, -1):
            suffixSum[i] = nums[i + 1] * suffixSum[i + 1]
        for i in range(size):
            ans.append(prefixSum[i] * suffixSum[i])
        return ans

File: Data Structures & Algorithms/find-duplicate-integer/submission-3.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int i = nums[0];
        while(i < nums.size()){
            if(nums[i] < 0)
                return i;
            nums[i] *= -1;
            i = abs(nums[i]);
        }
        return -1;
    }
};
